hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HStore.java

```diff
@@ -961,10 +961,10 @@ public List<KeyValueScanner> getScanners(boolean cacheBlocks, boolean usePread,
       storeFilesToScan = this.storeEngine.getStoreFileManager().getFilesForScan(startRow,
         includeStartRow, stopRow, includeStopRow);
       memStoreScanners = this.memstore.getScanners(readPt);
+      storeFilesToScan.stream().forEach(f->f.getReader().incrementRefCount());
```

Comment:
Is it possible that the storefiles are deleted before we increment ref count here?

Reply:
We are inside a storeEngine.readLock here. The delete code also is under the same lock, so if we reach this point, we are sure the files were not deleted yet.

Reply:
Then there is no problem for the old code either? Since we are under the lock? I do not fully get the point why here we could fix the problem? Mind adding more comment in the code? Thanks.

Reply:
OK, we will release the lock immediately after getting the file list, so when we actually get the store file scanner, the file may have already been archived.

I think a simple fix is to just move the getting scanner code into the read lock section? As here you only increase the lock if the reader is not null, it is still possible that some files are archived later?

Reply:
I think one problem of moving the getting scanner code into the read lock section is that in `StoreFileScanner.getScannersForStoreFiles `method ,  it would  call `HStoreFile.initReader` , which would make the `storeEngine.readLock `somewhat long time?   Another option is that since `StoreFileReader.refCount` comes from `StoreFileInfo.refCount`, we can increase the  StoreFileInfo.refCount by `HStoreFile.fileInfo`?

Reply:
> it would call HStoreFile.initReader , which would make the storeEngine.readLock locked somewhat long time? Another option is that since StoreFileReader.refCount comes from StoreFileInfo.refCount, we can increase the StoreFileInfo.refCount by HStoreFile.fileInfo?

Yeah, didn't want to move the SFR/SFS initialisation inside this read lock. I had updated the code to inc refCount via StoreFileInfo, as @comnetwork suggested.